workflow failure on doc building

Current actions for doc building are failing with the following error message:
```
Exception occurred:
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/flask/helpers.py", line 16, in <module>
    from werkzeug.urls import url_quote
ImportError: cannot import name 'url_quote' from 'werkzeug.urls' (/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/werkzeug/urls.py)
The full traceback has been saved in /tmp/sphinx-err-dybf5qvp.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
make: *** [Makefile:20: html] Error 2
Error: Process completed with exit code 2.
```

Failures seem to be systematic since at least today, the last successful doc building was [action 156](https://github.com/Helsinki-NLP/mammoth/actions/runs/6351988133), three days ago. Everything that was run today has been failing doc building,

Not entirely sure what's causing it; moreover I'm having a hard time reproducing it locally. My guess would be something related to dependencies.

--- a/.github/workflows/push.yml
+++ b/.github/workflows/push.yml
@@ -385,30 +385,3 @@ jobs:
 #          -rnn_size 16  -train_steps 20 \
 #          -update_vocab -reset_optim "states" \
 #          -train_from /tmp/lm.onmt.model_step_10.pt
-
-  build-docs:
-    runs-on: ubuntu-latest
-    steps:
-    - uses: actions/checkout@v2
-    - name: Set up Python 3.8
-      uses: actions/setup-python@v2
-      with:
-        python-version: 3.8
-    - name: Install dependencies
-      run: |
-        python -m pip install --upgrade pip
-        pip install --upgrade setuptools
-        pip install -e .
-        pip install -r docs/requirements.txt
-    - name: Build docs
-      run: |
-        set -e
-        # Check that docs are built without errors
-        cd docs/ && make html && cd ..
-    - name: Deploy docs
-      uses: JamesIves/github-pages-deploy-action@3.7.1
-      with:
-        GITHUB_TOKEN:  ${{ secrets.GITHUB_TOKEN }}
-        BRANCH: gh-pages
-        FOLDER: docs/build/html
-        CLEAN: true

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -1,6 +1,6 @@
 # Contributors
 
-OpenNMT-py is a community developed project and we love developer contributions.
+MAMMOTH, built on top of OpenNMT-py, is a community developed project and we love developer contributions.
 
 ## Guidelines
 Before sending a PR, please do this checklist first:

--- a/docs/Makefile
+++ b/docs/Makefile
@@ -4,7 +4,7 @@
 # You can set these variables from the command line.
 SPHINXOPTS    =
 SPHINXBUILD   = python3 -msphinx
-SPHINXPROJ    = OpenNMT-py
+SPHINXPROJ    = mammoth
 SOURCEDIR     = source
 BUILDDIR      = build
 

--- a/docs/source/CONTRIBUTING.md
+++ b/docs/source/CONTRIBUTING.md
@@ -1,6 +1,6 @@
 # Contributors
 
-OpenNMT-py is a community developed project and we love developer contributions.
+MAMMOTH, built on top of OpenNMT-py, is a community developed project and we love developer contributions.
 
 ## Guidelines
 Before sending a PR, please do this checklist first:

--- a/docs/source/FAQ.md
+++ b/docs/source/FAQ.md
@@ -1,0 +1,39 @@
+
+
+## About MAMMOTH
+### What are the key features and functionalities of the MAMMOTH? 
+
+MAMMOTH is a flexible for building modular sequence-to-sequence models. Its key features include modularity with different parameter sharing strategies and adapters, useful tools such as  automatic configure generation, and hyper-parameter optimization in scalable multilingual training. 
+
+### Are there any specific use cases or domains where MAMMOTH is particularly well-suited? 
+
+MAMMOTH can be use for training neural machine translation models from scratch using data in any domains. It is particularly well-suited for massively multilingual scenario. 
+
+### What unique approaches or algorithms does MAMMOTH uses?
+
+MAMMOTH uses the transformer-based sequence-to-sequence model. It adopts modular neural architectures with flexible sharing mechanisms.   
+
+### What is the development status of MAMMOTH, and are there plans for future enhancements or features?
+
+MAMMOTH is under active development now. We are testing MAMMOTH in massively multilingual setting, and plan to improve the modularity by adding support for different adapter networks, and consider sparse architectures with mixture-of-expert. 
+
+### How can users contribute to the development of MAMMOTH or report issues with the repository and documentation? 
+
+As a open-source project, we love contributions from the community. Users can refer to [`CONTRIBUTING.md`](./CONTRIBUTING.md) for details.
+
+
+
+## MAMMOTH and OpenNMT
+### What the relationship between MAMMOTH and the original OpenNMT project?
+
+MAMMOTH build upon OpenNMT-py - the PyTorch version of OpenNMT project. 
+
+### How does MAMMOTH leverage the capabilities of OpenNMT while introducing its own innovations?
+
+MAMMOTH chooses PyTorch-based OpenNMT, which allows it to fully leverage the capabilities of OpenNMT and PyTorch ecosystem to enable user friendly development.  
+
+### What are the modifications that MAMMOTH has made to OpenNMT-py? 
+
+MAMMOTH improves OpenNMT-py to enable more flexible modularity.  
+
+

--- a/docs/source/attention_bridges.md
+++ b/docs/source/attention_bridges.md
@@ -1,7 +1,7 @@
 
 # Attention Bridge
 
-The embeddings are generated through the self-attention mechanism ([Attention Bridge](./mammoth/modules/attention_bridge.py)) of the encoder and establish a connection with language-specific decoders that focus their attention on these embeddings. This is why they are referred to as 'bridges'. This architectural element serves to link the encoded information with the decoding process, enhancing the flow of information between different stages of language processing.
+The embeddings are generated through the self-attention mechanism of the encoder and establish a connection with language-specific decoders that focus their attention on these embeddings. This is why they are referred to as 'bridges'. This architectural element serves to link the encoded information with the decoding process, enhancing the flow of information between different stages of language processing.
 
 There are five types of attention mechanism implemented:
 
@@ -15,41 +15,13 @@ layer_type_to_cls = {
         }
 ```
 
-`intermediate_output` refers to the input of the attention bridge, specifically denoting the output that emerges from the encoder, i.e., $\mathbf{H}\in \mathbf{R}^{n\times d_h}$ `(batch, seq_len, hidden_dim)`. In other words, this is the transformed representation of the input data after it has undergone encoding processes. This intermediate output serves as the foundation for the subsequent attention bridge, facilitating the connection between the encoded information and the subsequent stages of language processing.
+`intermediate_output` refers to the input of the attention bridge, specifically denoting the output that emerges from the encoder, i.e., \\( \mathbf{H}\in \mathbf{R}^{n\times d_h} \\) `(batch, seq_len, hidden_dim)`. In other words, this is the transformed representation of the input data after it has undergone encoding processes. This intermediate output serves as the foundation for the subsequent attention bridge, facilitating the connection between the encoded information and the subsequent stages of language processing.
 
-<!-- TODO: double check this -->
-<aside>
-⚠️ The output of n-layered attention bridge is `[out, alphas]`, and each type of attention bridge outputs `[alphas, out]`
-</aside>
 
 ## LinAttentionBridgeLayer
 
 The attention bridge employed is based on the structured self-attention mechanism introduced by Lin et al. in their 2017 paper, accessible at [https://arxiv.org/abs/1703.03130](https://arxiv.org/abs/1703.03130). This mechanism is utilized to establish the connection between different components of the architecture.
-
-Mathematically, the attention bridge is represented as follows:
-
-Given an original input matrix $\mathbf{H} \in \mathbb{R}^{d_h \times n}$, where $d_h$ is the dimensionality of the input features and $n$ is the number of elements in the input sequence, the attention bridge generates a transformed output matrix $\mathbf{H}^\prime \in \mathbb{R}^{r \times n}$.
-
-Here, $\mathbf{W}_2$ is a weight matrix of dimensions $\mathbb{R}^{r \times d_a}$, where $r$ indicates the number of attention hops. The matrix $\mathbf{W}_1 \in \mathbb{R}^{d_a \times d_h}$ is another weight matrix of dimensions corresponding to the attention dimensions $d_a$. The term "multiple hops" implies the use of multiple attention hops to capture diverse relationships within the data.
-
-The mathematical operations are as follows:
-
-1. Compute the product of the weight matrices $\mathbf{W}_1$ and $\mathbf{H}^\top$:
-   $$\mathbf{W}_{1} \mathbf{H}^{\top}$$
-
-2. Apply the hyperbolic tangent (tanh) activation function to the result of the previous step:
-   $$\tanh \left(\mathbf{W}_{1} \mathbf{H}^{\top}\right)$$
-
-3. Multiply the outcome of the tanh operation by the weight matrix $\mathbf{W}_2$:
-   $$\mathbf{W}_{2} \tanh \left(\mathbf{W}_{1} \mathbf{H}^{\top}\right)$$
-
-4. Apply the softmax function to normalize the weights along the specified dimension:
-   $$\operatorname{Softmax}\left(\mathbf{W}_{2} \tanh \left(\mathbf{W}_{1} \mathbf{H}^{\top}\right)\right)$$
-
-5. Finally, perform element-wise multiplication with the original input matrix $\mathbf{H}$:
-   $$\operatorname{Softmax}\left(\mathbf{W}_{2} \tanh \left(\mathbf{W}_{1} \mathbf{H}^{\top}\right)\right) \mathbf{H}$$
-
-The resulting matrix $\mathbf{H}^\prime$ encapsulates the enhanced relationships and features that are established through the structured self-attention mechanism. This attention bridge allows the encoded information to be channeled effectively into subsequent language-specific decoding processes, contributing to the overall performance of the architecture.
+This attention bridge allows the encoded information to be channeled effectively into subsequent language-specific decoding processes, contributing to the overall performance of the architecture.
 
 ## PerceiverAttentionBridgeLayer
 
@@ -71,13 +43,13 @@ The `PerceiverAttentionBridgeLayer` involves a multi-headed dot product self-att
 
 The process described involves dot product self-attention. The steps are as follows:
 
-1. **Input Transformation**: Given an input matrix $\mathbf{H} \in \mathbb{R}^{d_h \times n}$, two sets of learned weight matrices are used to transform the input. These weight matrices are $\mathbf{W}_1 \in \mathbb{R}^{d_h \times d_a}$ and $\mathbf{W}_2 \in \mathbb{R}^{d_h \times d_a}$. The multiplication of $\mathbf{H}$ with $\mathbf{W}_1$ and $\mathbf{W}_2$ produces matrices $\mathbf{V}$ and $\mathbf{K}$, respectively:
+1. **Input Transformation**: Given an input matrix \\(\mathbf{H} \in \mathbb{R}^{d_h \times n}\\), two sets of learned weight matrices are used to transform the input. These weight matrices are \\( \mathbf{W}_1 \in \mathbb{R}^{d_h \times d_a}\\) and \\( \mathbf{W}_2 \in \mathbb{R}^{d_h \times d_a}\\). The multiplication of \\(\mathbf{H}\\) with \\(\mathbf{W}_1\\) and \\( \mathbf{W}_2\\) produces matrices \\( \mathbf{V} \\) and \\( \mathbf{K}\\), respectively:
 
-   - $\mathbf{V} = \mathbf{H} \mathbf{W}_1$
-   - $\mathbf{K} = \mathbf{H} \mathbf{W}_2$
+   - \\( \mathbf{V} = \mathbf{H} \mathbf{W}_1\\)
+   - \\( \mathbf{K} = \mathbf{H} \mathbf{W}_2\\)
 
-2. **Attention Calculation**: The core attention calculation involves three matrices: $\mathbf{Q} \in \mathbb{R}^{d_h \times n}$, $\mathbf{K}$ (calculated previously), and $\mathbf{V}$ (calculated previously). The dot product of $\mathbf{Q}$ and $\mathbf{K}^\top$ is divided by the square root of the dimensionality of the input features ($\sqrt{d_h}$).
-The final attended output is calculated by multiplying the attention weights with the $\mathbf{V}$ matrix: $\mathbf{H}^\prime = \operatorname{Softmax}(\frac{\mathbf{Q}\mathbf{K}^\top}{\sqrt{d_h}})\mathbf{V}$
+2. **Attention Calculation**: The core attention calculation involves three matrices: \\( \mathbf{Q} \in \mathbb{R}^{d_h \times n}\\), \\( \mathbf{K} \\) (calculated previously), and \\( \mathbf{V}\\) (calculated previously). The dot product of \\( \mathbf{Q}\\) and \\(\mathbf{K}^\top\\) is divided by the square root of the dimensionality of the input features (\\( \sqrt{d_h}\\)).
+The final attended output is calculated by multiplying the attention weights with the \\( \mathbf{V} \\) matrix: \\( \mathbf{H}^\prime = \operatorname{Softmax}(\frac{\mathbf{Q}\mathbf{K}^\top}{\sqrt{d_h}})\mathbf{V}\\)
 
 
 ## TransformerAttentionBridgeLayer

--- a/docs/source/config_config.md
+++ b/docs/source/config_config.md
@@ -1,20 +1,21 @@
 # Config-config tool
 
-A meta-configuration tool, or config generator for FoTran OpenNMT.
+A meta-configuration tool, or config generator for MAMMOTH.
 
-The FoTran OpenNMT configuration options have become unwieldly as complexity has increased.
+The MAMMOTH configuration options have become unwieldy as complexity has increased.
 Especially the introduction of LayerStacks (the mechanism for dividing encoders and decoders into several subcomponents with different parameter sharing) and Adapters has made specifying parameters on the command line into a total nightmare, and even writing yaml configs by hand is cumbersome.
 Other functionality that is cumbersome to specify by hand includes:
 
-  - Node and GPU assigments for massively multilingual models,
+  - Node and GPU assignments for massively multilingual models,
   - Weights and curricula (starting step for specific tasks),
   - Parameter sharing groups based on language similarity.
 
 To ease the creation of configs, the config-config tool reads in a human-writable configuration template, and computes the specific values expected by OpenNMT, writing them out as a less-readable yaml file.
 
 ## Command
-
-python3 OpenNMT-py-v2/tools/config\_config.py config\_all --in\_config input.yaml --out\_config output.yaml
+```bash
+python3 mammoth/tools/config_config.py config_all --in_config input.yaml --out_config output.yaml
+```
 
 ## Inputs
 
@@ -25,7 +26,7 @@ The primary input is a yaml file, which contains two types of parameters
 
 ### Input yaml
 
-See the example in `OpenNMT-py-v2/examples/config_config.yaml`
+See the example in `mammoth/examples/config_config.yaml`
 
 The meta-parameters under the `config_config` key:
 
@@ -75,13 +76,13 @@ The number of language groups to create when clustering.
 
 If set to `True`, use corpus weights based on temperature-adjusted corpus size.
 
-Note that the actual weight is proportional to the weights of the the other tasks assinged to the same GPU. E.g. if only one task is assigned to a GPU, it will receive 100% weight regardless of what the computed weight is.
+Note that the actual weight is proportional to the weights of the the other tasks assigned to the same GPU. E.g. if only one task is assigned to a GPU, it will receive 100% weight regardless of what the computed weight is.
 
 #### `use_introduce_at_training_step`
 
 If set to `True`, use a curriculum introducing corpora based on temperature-adjusted corpus size.
 
-Note that if both `use_weight` and `use_introduce_at_training_step` are specificed, the weight is distributed to the two according to the square root, so that when both of them are applied (multiplicatively), the desired weight is achieved.
+Note that if both `use_weight` and `use_introduce_at_training_step` are specified, the weight is distributed to the two according to the square root, so that when both of them are applied (multiplicatively), the desired weight is achieved.
 
 Note that high-resource language pairs (would train for over 75% of the training time) all start at 0. This avoids starting training with only one GPU doing work, while the other GPUs are idle waiting for their LPs to start.
 
@@ -124,7 +125,7 @@ Note that it is possible to have target-language-dependent components in the enc
 
 #### `n_nodes` and `n_gpus_per_node`
 
-The number of nodes and GPUs, for assigment of tasks to devices.
+The number of nodes and GPUs, for assignment of tasks to devices.
 Note that you also need to separately specify this information to slurm.
 
 #### Other top-level keys than `config_config`
@@ -141,13 +142,13 @@ The key `adapters.encoder.{adapter_name}.ids` takes one of 3 values:
 
 ### Distance matrix
 
-See the example distance matrix in `OpenNMT-py-v2/examples/config_config.distance.csv`.
+See the example distance matrix in `mammoth/examples/config_config.distance.csv`.
 
 The distance matrix is given as a csv file, with a column `lang` and one column per language.
 There should be one row per language, with the language code in the first `lang` column, followed by a float giving the distance to the language specified by the column.
 The rows should appear in the same order as the columns.
 This means that the matrix must be square and symmetrical.
-The upper and lower triange are redundant, but both must be given.
+The upper and lower triangle are redundant, but both must be given.
 
 Note that the values are distances: the distance of a language to itself (the diagonal) should be 0.
 
@@ -168,8 +169,9 @@ The actual corpus files are used in two ways:
 Because of this, you need to run `config_config` so that it can access the corpora using the specified `src_path` and `tgt_path`.
 
 ## Usage
-
-    python OpenNMT-py-v2/tools/config_config.py config_all --in_config path/to/input.yaml --out_config path/to/output.yaml
+```
+python mammoth/tools/config_config.py config_all --in_config path/to/input.yaml --out_config path/to/output.yaml
+```
 
 ### Stages
 

--- a/docs/source/index.rst
+++ b/docs/source/index.rst
@@ -7,7 +7,6 @@ Contents
 
       main.md
       quickstart.md
-      install.md
       CONTRIBUTING.md
       ref.rst
 
@@ -21,6 +20,7 @@ Contents
       :caption: MAMMOTH features
       :maxdepth: 2
 
+      modular_model.md
       config_config.md
       attention_bridges.md
 
@@ -28,6 +28,7 @@ Contents
       :caption: Tutorials
       :maxdepth: 2
 
+      install.md
       prepare_data.md
       examples/Translation.md
 

--- a/docs/source/index.rst
+++ b/docs/source/index.rst
@@ -12,6 +12,12 @@ Contents
       ref.rst
 
 .. toctree::
+      :caption: Frequently Asked Questions
+      :maxdepth: 2
+
+      FAQ.md
+
+.. toctree::
       :caption: MAMMOTH features
       :maxdepth: 2
 

--- a/docs/source/install.md
+++ b/docs/source/install.md
@@ -11,7 +11,7 @@ In the login node, create a directory in the `projappl` linked to our project to
 # where to install the necessary python packages
 ENV_DIR="/projappl/project_2005099/test"
 # where the codebase was copied to
-CODE_DIR="/scratch/project_2005099/path/to/OpenNMT-py-v2"
+CODE_DIR="/scratch/project_2005099/path/to/mammoth"
 
 # set up variables & modules
 module load pytorch
@@ -32,7 +32,7 @@ In slurm job scripts, update environment variables to get python to run your cod
 
 ```
 ENV_DIR="/projappl/project_2005099/test"
-CODE_DIR="/scratch/project_2005099/path/to/OpenNMT-py-v2"
+CODE_DIR="/scratch/project_2005099/path/to/mammoth"
 
 module load pytorch
 export PYTHONUSERBASE=$ENV_DIR
@@ -63,7 +63,7 @@ srun python3 -u $CODE_DIR/train.py ...
 5. Install mammoth 
     
     ```bash
-    cd /pfs/lustrep1/projappl/${PROJECT}/${USER}/OpenNMT-py-v2
+    cd /pfs/lustrep1/projappl/${PROJECT}/${USER}/mammoth
     pip3 install -e .
     pip3 install sentencepiece==0.1.97 sacrebleu==2.3.1
     ```

--- a/docs/source/install.md
+++ b/docs/source/install.md
@@ -1,17 +1,21 @@
 
 # Installation
 
-## Puhti / Mahti
+Installing from source usually works. Here, we provide two examples that install MAMMOTH in Nvidia and AMD platforms. 
+
+## Puhti / Mahti 
+
+This first example install MAMMOTH on clusters ([Puhti/Mahti](https://docs.csc.fi/computing/)) with Nvidia GPUs.
 
 ### Install
 In the login node, create a directory in the `projappl` linked to our project to host the shared python dependencies, and install the code base & dependencies there:
 
-<!-- TODO: modify project name  -->
 ```
 # where to install the necessary python packages
-ENV_DIR="/projappl/project_2005099/test"
+PROJECT=<your_project_name_is_your_account>
+ENV_DIR="/projappl/${PROJECT}/test"
 # where the codebase was copied to
-CODE_DIR="/scratch/project_2005099/path/to/mammoth"
+CODE_DIR="/scratch/${PROJECT}/path/to/mammoth"
 
 # set up variables & modules
 module load pytorch
@@ -31,8 +35,8 @@ chmod -R 777 $CODE_DIR
 In slurm job scripts, update environment variables to get python to run your code properly:
 
 ```
-ENV_DIR="/projappl/project_2005099/test"
-CODE_DIR="/scratch/project_2005099/path/to/mammoth"
+ENV_DIR="/projappl/${PROJECT}/test"
+CODE_DIR="/scratch/${PROJECT}/path/to/mammoth"
 
 module load pytorch
 export PYTHONUSERBASE=$ENV_DIR
@@ -45,8 +49,12 @@ srun python3 -u $CODE_DIR/train.py ...
 
 ## LUMI
 
+The GPU partition of [LUMI supercomputer](https://lumi-supercomputer.github.io) is AMD-based. 
+This example uses python virtual environment without container. But as recommended by the LUST team of LUMI, better to use container. An instruction to install MAMMOTH under Singularity container will release in the future.
+<!-- TODO -->
+
 ### Install 
-1. start an interactive session `srun --account="$PROJECT" --partition=dev-g --ntasks=1 --gres=gpu:mi250:1 --time=2:00:00 --mem=25G --pty bash`
+1. Start an interactive session `srun --account="$PROJECT" --partition=dev-g --ntasks=1 --gres=gpu:mi250:1 --time=2:00:00 --mem=25G --pty bash`
 2. Load modules: 
     
     ```bash
@@ -69,5 +77,8 @@ srun python3 -u $CODE_DIR/train.py ...
     ```
 
 ### Run
-
-TODO
+You can train in slurm job scripts, such as
+```bash
+srun python -u $CODE_DIR/train.py ...
+```
+For more details, we refer to the tutorial section. 

--- a/docs/source/main.md
+++ b/docs/source/main.md
@@ -24,6 +24,9 @@ This project is based on [OpenNMT V2](https://opennmt.net).
 When using OpenNMT-py for research please cite the
 [OpenNMT technical report](https://doi.org/10.18653/v1/P17-4012)
 
+We published [FoTraNMT](https://github.com/Helsinki-NLP/FoTraNMT) the ancestor of MAMMOTH, when using MAMMOTH for research, please cite [Boggia et al. (2023)](https://aclanthology.org/2023.nodalida-1.24).
+
+
 ```
 @inproceedings{opennmt,
   author    = {Guillaume Klein and
@@ -36,5 +39,23 @@ When using OpenNMT-py for research please cite the
   year      = {2017},
   url       = {https://doi.org/10.18653/v1/P17-4012},
   doi       = {10.18653/v1/P17-4012}
+}
+
+@inproceedings{boggia-etal-2023-dozens,
+    title = "Dozens of Translation Directions or Millions of Shared Parameters? Comparing Two Types of Multilinguality in Modular Machine Translation",
+    author = {Boggia, Michele  and
+      Gr{\"o}nroos, Stig-Arne  and
+      Loppi, Niki  and
+      Mickus, Timothee  and
+      Raganato, Alessandro  and
+      Tiedemann, J{\"o}rg  and
+      V{\'a}zquez, Ra{\'u}l},
+    booktitle = "Proceedings of the 24th Nordic Conference on Computational Linguistics (NoDaLiDa)",
+    month = may,
+    year = "2023",
+    address = "T{\'o}rshavn, Faroe Islands",
+    publisher = "University of Tartu Library",
+    url = "https://aclanthology.org/2023.nodalida-1.24",
+    pages = "238--247"
 }
 ```

--- a/docs/source/modular_model.md
+++ b/docs/source/modular_model.md
@@ -1,0 +1,26 @@
+# Modular Model
+
+MAMMOTH is a modular framework for machine translation that offers flexibility in designing different sharing schemes for its modules. Let's break down the key aspects and features mentioned:
+
+1. **Depth of Encoder & Decoder**:
+    - **Balanced**: The encoder and decoder have the same depths.
+    - **Deep-enc-shallow-dec**: The encoder is deep, while the decoder is relatively shallow.
+
+2. **Target Language Selection Token**:
+    - There is an option to include or exclude a target language selection token in the architecture.
+
+3. **Layerwise Parameter Sharing Schemes**:
+    - **Fully Shared Encoder and Fully Shared Decoder (Baseline)**: Both the encoder and decoder have shared parameters, meaning they are common across all languages or translation pairs.
+    - **Fully Shared Encoder and Target-Specific Decoder**: The encoder is shared, but each target language has its own decoder.
+    - **Apple-Style Learning Language-specific Layers**: The encoder contains both source and target-specific layers, which suggests a more intricate sharing pattern.
+    - **Neural Architecture Search**: Potentially, a method to automatically discover optimal sharing patterns.
+    - **Adapter-Like Low-Rank Residual Layers**: This implies the use of adapter-like layers with low-rank residual connections, which can facilitate adaptation for different languages or translation tasks.
+
+4. **Groups for Groupwise Shared Parameters**:
+    - **Not Used (Only Fully Shared and Language-Specific)**: In this case, there are no specific parameter sharing groups beyond fully shared and language-specific ones.
+    - **Phylogenetic**: Parameters could be shared among languages that are phylogenetically related, meaning they share a common ancestry.
+    - **Clustering Based on Typological Database**: Sharing could be determined based on linguistic typological features or characteristics.
+    - **Clustering Based on Language Embeddings**: This could involve sharing parameters based on the embeddings of languages in a common vector space.
+
+5. **Subword Vocabularies**:
+    - **Fully Shared vs Language-Specific**: The system allows for a choice between using a fully shared subword vocabulary for all languages or using language-specific subword vocabularies, which can be beneficial for languages with unique linguistic characteristics.

--- a/docs/source/refs.bib
+++ b/docs/source/refs.bib
@@ -1,3 +1,38 @@
+@inproceedings{boggia-etal-2023-dozens,
+    title = "Dozens of Translation Directions or Millions of Shared Parameters? Comparing Two Types of Multilinguality in Modular Machine Translation",
+    author = {Boggia, Michele  and
+      Gr{\"o}nroos, Stig-Arne  and
+      Loppi, Niki  and
+      Mickus, Timothee  and
+      Raganato, Alessandro  and
+      Tiedemann, J{\"o}rg  and
+      V{\'a}zquez, Ra{\'u}l},
+    booktitle = "Proceedings of the 24th Nordic Conference on Computational Linguistics (NoDaLiDa)",
+    month = may,
+    year = "2023",
+    address = "T{\'o}rshavn, Faroe Islands",
+    publisher = "University of Tartu Library",
+    url = "https://aclanthology.org/2023.nodalida-1.24",
+    pages = "238--247"
+}
+
+@inproceedings{pires-etal-2023-learning,
+    title = "Learning Language-Specific Layers for Multilingual Machine Translation",
+    author = "Pires, Telmo  and
+      Schmidt, Robin  and
+      Liao, Yi-Hsiu  and
+      Peitz, Stephan",
+    booktitle = "Proceedings of the 61st Annual Meeting of the Association for Computational Linguistics (Volume 1: Long Papers)",
+    month = jul,
+    year = "2023",
+    address = "Toronto, Canada",
+    publisher = "Association for Computational Linguistics",
+    url = "https://aclanthology.org/2023.acl-long.825",
+    doi = "10.18653/v1/2023.acl-long.825",
+    pages = "14767--14783",
+}
+
+
 @article{DBLP:journals/corr/LiuL17d,
   author    = {Yang Liu and
                  Mirella Lapata},

